synapse/handlers/federation.py

```diff
@@ -1695,22 +1695,84 @@ async def on_make_join_request(
             },
         )
 
+        # If the room is a restricted room the auth event ID calculation is a bit
+        # more complicated than the generic methods handle.
+        additional_auth_ids = None
+        current_state_ids = await self.store.get_current_state_ids(room_id)
+        if await self._event_auth_handler.has_restricted_join_rules(
+            current_state_ids, room_version
+        ):
+            additional_auth_ids = [
+                await self._get_user_event_which_could_invite(
+                    room_id, current_state_ids
+                )
+            ]
+
         try:
             event, context = await self.event_creation_handler.create_new_client_event(
-                builder=builder
+                builder=builder, additional_auth_event_ids=additional_auth_ids
             )
         except SynapseError as e:
             logger.warning("Failed to create join to %s because %s", room_id, e)
             raise
 
+        # Ensure the user can even join the room.
+        await self._check_join_restrictions(context, event)
+
         # The remote hasn't signed it yet, obviously. We'll do the full checks
         # when we get the event back in `on_send_join_request`
         await self.auth.check_from_context(
-            room_version, event, context, do_sig_check=False
+            room_version.identifier, event, context, do_sig_check=False
         )
 
         return event
 
+    async def _get_user_event_which_could_invite(
```

Comment:
I'm surprised I don't seem to need to call this in `send_join`, but I might be getting "lucky" in my testing that the room state hasn't changed between `make_join` and `send_join`?